File: Solutions/735_Asteroid_Collision/735_Asteroid_Collision.cpp
//Third fastest
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> theStack;
        for(int i =0 ; i < asteroids.size(); ++i)
        {
           if(theStack.size() == 0)
           {
                theStack.push(asteroids[i]);
                continue;
           }
            
            // stack top explode only
            while(theStack.size() > 0 && asteroids[i] < 0 && theStack.top() > 0 && abs(asteroids[i]) > abs(theStack.top()))
            { 
                // stack top explode
                theStack.pop();    
                // then keep checking asteroids[i] with next top of the stack
            }
            
            //asteroids[i] explode only
            if(theStack.size() > 0 && asteroids[i] < 0 && theStack.top() > 0  && abs(asteroids[i]) < abs(theStack.top()))
            {
                // continue to next i
                continue;
            }
            
            //both explode 
            if(theStack.size() > 0 && asteroids[i] < 0 && theStack.top() > 0  && abs(asteroids[i]) == abs(theStack.top()))
            {
                // stack top explode
                theStack.pop();  
                
                // continue to next i
                continue;
            }
            
            // other cases, the same direction 
            theStack.push(asteroids[i]);
            
        }
      vector<int> res;
       while(theStack.size() > 0)
       {
           res.push_back(theStack.top());
           theStack.pop();
       }
        
        reverse(res.begin(), res.end());
        return res;
        
    }
};
//Second fastest
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> s;
        for(int i=0;i<asteroids.size();i++){
            bool flag=true;
            while(s.size()&&s.top()>0&&asteroids[i]<0){
                if(abs(s.top())>abs(asteroids[i])){
                    flag=false;
                    break;
                }
                else if(abs(s.top())==abs(asteroids[i])){
                    s.pop();
                    flag=false;
                    break;
                }
                else s.pop();
            }
            if(flag)
                s.push(asteroids[i]);
        }
        vector<int> v;
        while(s.size()){
            v.push_back(s.top());
            s.pop();
        }
        reverse(v.begin(),v.end());
        return v;
    }
};

//Fastest
bool same(int a, int b) {
    return (a > 0 && b > 0) || (a < 0 && b < 0);
}

void print_deque(deque<int>& d) {
    cout << "d=";
    for (int i : d) {
        cout << i << " ";
    }
    cout << endl;
}

class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        
        deque<int> result;
        
        for (int i = 0; i < (int)asteroids.size(); ++i) {
            //cout << "\n\nstart!\nastr=" << asteroids[i] << endl;
            //print_deque(result);
            
            if (result.empty()) {
                //cout << "just add" << endl;
                result.push_back(asteroids[i]);
                continue;
            }
            else if (same(asteroids[i], result.back())) { // the same direction 
                result.push_back(asteroids[i]);
                //cout << "add - same dir" << endl;
            }
            else if (result.back() < 0 && asteroids[i] > 0) {
                //cout << "just add - they won't meet" << endl;
                result.push_back(asteroids[i]);
            }
            else if (abs(result.back()) > abs(asteroids[i])) { // different directions, but explosion
                // do nothing
                //cout << "do nothing - our element is bigger" << endl;
            }
            else if (abs(result.back()) == abs(asteroids[i])) {
                result.pop_back();
                //cout << "eliminate both - same value" << endl;
            }
            else {
                result.pop_back();
                --i;
                //cout << "remove from result and give one more change" << endl;
            }
        }
        
        return vector(result.begin(), result.end());
        
    }
};

File: Solutions/735_Asteroid_Collision/735_Asteroid_Collision.js
//Third fastest
/**
 * @param {number[]} asteroids
 * @return {number[]}
 */
var asteroidCollision = function(asteroids) {
    if(asteroids.length <= 1) {
        return asteroids;
    }
    let sign = Math.sign(asteroids[0]) // flag value for positive
    let result = [];
    let i = 0;
    while(i < asteroids.length) {
        if (Math.sign(asteroids[i]) != sign && sign > 0) {
            if (result.length > 0) {
                let len = result.length -1;
                if (result[len] < Math.abs(asteroids[i])) {
                    result.pop();
                    sign = Math.sign(result[len-1]);
                } else if(result[len] == Math.abs(asteroids[i])){
                    result.pop();
                    sign = Math.sign(result[len-1]);
                    i++;
                } else {
                    i++;
                }    
            } else {
                result.push(asteroids[i]);
                sign = Math.sign(asteroids[i]);
                i++;
            }
        } else {
            result.push(asteroids[i]);
            sign = Math.sign(asteroids[i]);
            i++;
        }

    }

    return result;
    
};

//Second fastest
/**
 * @param {number[]} asteroids
 * @return {number[]}
 */
var asteroidCollision = function(asteroids) {
    let stack = [asteroids[0]]
    for (let i = 1; i < asteroids.length; i++) {
        if ( stack.length === 0 || asteroids[i] > 0 ) {
            stack.push(asteroids[i]);
        } else {
            while (stack.length) {
                let ele = stack.pop();
                
                if (ele < 0) {
                    stack.push(ele);
                    stack.push(asteroids[i]);
                    break;
                } else if (ele === Math.abs(asteroids[i])) {
                    break;
                } else if (ele > Math.abs(asteroids[i])) {
                    stack.push(ele);
                    break;
                } else {
                    if (stack.length === 0) {
                        stack.push(asteroids[i]);
                        break;
                    }
                }
            }
        }
    }
    return stack;
};

// [5, 10, -5]
// [8, -8]
// [10, 2, -5]
// [-2, -1, 1, 2]

//Fastest
var asteroidCollision = function(asteroids) {
  let result = [];
  let i = 0;
  
  while (i < asteroids.length) {
    let lastElem = result[result.length - 1];
    let sum = lastElem + asteroids[i];
    if (result.length === 0 || lastElem < 0) {
      result.push(asteroids[i]);
      i++;
    } else if (lastElem > 0 && asteroids[i] < 0) {
      if (sum === 0) {
        result.pop();
        i++;
      } else if (sum > 0) {
        i++;
      } else if (sum < 0) {
        result.pop();
      }
    } else {
      result.push(asteroids[i]);
      i++;
    }
  }
  return result;
};